File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/lc-all-solutions/206.reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

c_ Solution(o..):
  ___ reverseList  root):
    __ n.. root o. n.. root.next:
      r.. root

    ret = reverseList(root.next)
    root.next.next = root
    root.next = N..
    r.. ret

  ___ _reverseList  head):
    pre = N..
    cur = head
    w.... cur:
      tmp = cur.next
      cur.next = pre
      pre = cur
      cur = tmp
    r.. pre

  # iteratively as queue head inserting
  ___ __reverseList  head):
    """
    :type head: ListNode
    :rtype: ListNode
    """
    dHead = dummy = ListNode(-1)
    p = head
    w.... p:
      tmp = dummy.next
      dummy.next = p
      p = p.next
      dummy.next.next = tmp
    r.. dHead.next

  # easily leads to a circle. Remove current node's next after recursive call.
  ___ ___reverseList  head):
    newHead = N..

    ___ rec(head):
      __ n.. head:
        r.. head
      p = rec(head.next)
      head.next = N..
      __ p:
        p.next = head
      ____:
        newHead = head
      r.. head

    rec(head)
    r.. newHead

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Array/ReverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// 翻转链表的思路
// 把head取出来，把head.next取出来，head.next.next也取出来。
// head.next = head
// 新的(head.next和head)与head.next.next进行新一轮交换。
// 直至空。
// PS，链表操作好麻烦。
 var reverseList = function(head) {

    let newHead = null

    // 先不判断进行一次交换
    let nHead = head
    __ (!nHead) {
        r_ null
    }
    let next = nHead.next
    __ (!next) {
        r_ nHead
    }
    let lNext = next.next

    next.next = nHead
    nHead.next = null

    newHead = next

    _____ (nHead && next && next.next) {
        nHead = lNext
        __ (!nHead) {
            r_ newHead
        }

        next = nHead.next

        __ (!next) {
            nHead.next = newHead
            r_ nHead
        }
        lNext = next.next

        next.next = nHead
        nHead.next = newHead

        newHead = next

    }
};